### ISSUE TITLE
DID codec behavior for DIDs not in DidConfig

### ISSUE THREAD

**Issue by lapo4719** [pylessard/python-udsoncan#185]
I want to read DTC snapshots without adding a DID codec to the DID codec dictionary in the userconfig's.
I want to provide/define a default codec parsing behavior where it will just spit out the raw bytes to the response.service_data() when it can do nothing better, without it failing on me like shown in the error message:

Context:
I am receiving DIDs that I don't have codec entries for (this is a problem I could avoid on my end and this all wouldn't be a problem, but I'm seeking to handle this situation with raw bytes data, rather than the exception that occurs.)

this is where I'm adding my current DID# to Codec dictionary, but I don't have DID entries for the DIDs I'm receiving.
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/e73edb3d-985b-48a6-aae9-cf7563da39fa)

This is where I'm reading DTC data, where the library will try to decode the received DIDs
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/be4b74ae-049d-41d2-8b1b-e1c57a51e008)

`  File "C:\Git\diagnostics_tool\lib_diagnostic_tool\uds_utils\uds_driver.py", line 666, in request_read_dtc_snapshot_info
    dtc_snapshot_response: Response = self.uds_client.get_dtc_snapshot_by_dtc_number(dtc=dtc_id, record_number=snapshot_record_depth)
                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\client.py", line 1413, in get_dtc_snapshot_by_dtc_number
    return self.read_dtc_information(services.ReadDTCInformation.Subfunction.reportDTCSnapshotRecordByDTCNumber, dtc=dtc, snapshot_record_number=record_number)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\client.py", line 128, in decorated
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\client.py", line 1626, in read_dtc_information
    raise error
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\client.py", line 1610, in read_dtc_information
    services.ReadDTCInformation.interpret_response(response, **response_params)
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\services\ReadDTCInformation.py", line 564, in interpret_response
    didconfig = ServiceHelper.check_did_config(did, didconfig)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Git\diagnostics_tool\venv\Lib\site-packages\udsoncan\services\__init__.py", line 129, in check_did_config
    raise ConfigError(did, msg='Actual data identifier configuration contains no definition for data identifier 0x%04x' % did)
udsoncan.exceptions.ConfigError: Actual data identifier configuration contains no definition for data identifier 0xc422`

And this is the error for not have a configuration for this DID.

**Possible Solution**
There is way to add a feature for the user to control the default codec behavior on their end:
I can use a `defaultdict ` instead of a regular dictionary, and the user can add their own default DidCodec like below
`did_nums_to_codecs: defaultdict[int, DidCodec] = defaultdict(partial(DidCodec, 64))`

`print (did_nums_to_codecs['missing_codec'])`
`result >> <udsoncan.DidCodec object at 0x000002A1FF553550>`

but, the currently implemented way of checking that the did exists prevents this from working:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/be1f7e34-7bbf-4dc0-9b7e-4993f253aac2)
^ this `if key in dict` check won't kick-off the `defaultdict`'s factory. If the check in the library is instead replaced with `if not didconfig.get(did):` then a default can be provided even for DIDs that aren't known.

Could this check be changed in the library?

Thank you greatly for your work

**Comment by pylessard** [comment]
I understand. Well explained and reasonable request. Expect this to be fixed sooner than later. Cheers

**Comment by pylessard** [comment]
Can you give a try to this branch? 
https://github.com/pylessard/python-udsoncan/tree/handle-default-did

**Comment by lapo4719** [comment]
Thank you for the response!

I made a small mistake, defaultdict does not trigger its .default_factory on a mydict.get(key) call, it triggers it only when we use the brackets:

![image](https://github.com/pylessard/python-udsoncan/assets/48290683/ba63c6bd-56a3-4427-ac9c-fe794aade328)


The __missing__(key) dunder that defaultdict overrides has the same interface as `value = mydictionary[key]`, which is that if it fails it throws a KeyError. Unlike .get(), which returns None by default or a user provided `default`.

Can you switch it to use the following:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/51abf521-308b-4dd7-9be9-251bbe8eed5e)
https://docs.python.org/3/library/collections.html#collections.defaultdict
https://realpython.com/python-defaultdict/
^ This does work! At least in creating this small behavior and creates a new dict entry (it went from 20 ->21)

**Next Steps:**
But now I must decide how best to construct a default codec. it also appears that just using Partial(DidCodec, struct_string) means that I need one struct_string that always works. And struct.pack/unpack doesn't allow variable length, **so lack of a known/dynamic length feels like a fundamental problem. But unless you can magically detect DID length at run time, I think this is my problem. lets keep going and see if it can partially work anyway:**

using `did_config = defaultdict(partial(DidCodec, '10s') # which should just be 10 characters, idk it might get chopped off`

Sadly
The `.get_dtc_snapshot_by_dtc_number()` call doesn't parse the service data as expected, and drops the data. (I wasn't seeing any of these did's fail to fill the service_data[0] before when I provide a valid DID definition.)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/a146c0c3-fc3d-431d-9851-eb7bdf31ba7c)

Digging into it, reading the data is fine, it has a reasonable payload:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/2ab5daee-7ef1-400c-991f-42006b015a8c)

But sometimes .service_data didn't get filled with the dtcs[]. I not sure why or how all this works. 
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/82c59d6a-717c-4072-9f6f-25bdd822759f)

The data gets properly parsed as a snapshot record and a dtc gets appended to dtcs[]:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/754591b7-4aed-4d10-aae4-e7c913c791c2)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/2a1c5a3c-b808-4662-afab-d6e084964799)

On some of the get_dtc_snapshot_by_dtc_number() calls it doesn't fill the dtcs[] with anything, and it fails. I notice this is always when the data is clearly of a good size, having snapshots itn it
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/a55a746c-9245-4c0f-8cc8-d73adab724e3)

![image](https://github.com/pylessard/python-udsoncan/assets/48290683/4e078972-3fed-460c-a304-92e5c3bacb61)
^ This fails.

Okay posting this now. Gonna keep seeing why this happens..

**Comment by lapo4719** [comment]
**Next Steps:**
But now I must decide how best to construct a default codec. it also appears that just using Partial(DidCodec, struct_string) means that I need one struct_string that always works. And struct.pack/unpack doesn't allow variable length, **so lack of a known/dynamic length feels like a fundamental problem. But unless you can magically detect DID length at run time, I think this is my problem. lets keep going and see if it can partially work anyway:**

using `did_config = defaultdict(partial(DidCodec, '10s') # which should just be 10 characters, idk it might get chopped off`

Sadly
The `.get_dtc_snapshot_by_dtc_number()` call doesn't parse the service data as expected, and drops the data. (I wasn't seeing any of these did's fail to fill the service_data[0] before when I provide a valid DID definition.)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/a146c0c3-fc3d-431d-9851-eb7bdf31ba7c)

Digging into it, reading the data is fine, it has a reasonable payload:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/2ab5daee-7ef1-400c-991f-42006b015a8c)

But sometimes .service_data didn't get filled with the dtcs[]. I not sure why or how all this works. 
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/82c59d6a-717c-4072-9f6f-25bdd822759f)

The data gets properly parsed as a snapshot record and a dtc gets appended to dtcs[]:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/754591b7-4aed-4d10-aae4-e7c913c791c2)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/2a1c5a3c-b808-4662-afab-d6e084964799)

On some of the get_dtc_snapshot_by_dtc_number() calls it doesn't fill the dtcs[] with anything, and it fails.

**EDIT: I took more care and redid the screenshots:**

This dtc data works and gets populated
(library parsing of data)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/7707b3b9-e685-4136-ba0f-e925589a4b94)
(user code reading it after)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/0fe390be-f4b5-438b-b56f-eefe58e1095f)

This dtc data doesn't get populated and fails
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/67c7f390-29da-4742-b192-ef16d5f2ccd1)

**Comment by pylessard** [comment]
I can add a "default_codec" parameter if that helps.
You can also implement a dict-like object that defines __contains__ and __getitem__  I think that will go through.

Let me know what you think

**Comment by lapo4719** [comment]
I think that might be better

**Comment by pylessard** [comment]
Which of the 2 options?

**Comment by lapo4719** [comment]
Sorry I misread. I think the first idea, a default_codec, would be good.

**Comment by pylessard** [comment]
Will have a look at this soon.

**Comment by lapo4719** [comment]
Is there a way to define a DidCodec that has arbitrary length or can figure out the length of the raw payload? I'm not an expert at UDS services, and I'm not sure it allows for it, but that would be very nice behavior.

**Comment by pylessard** [comment]
The way UDS is designed, you need to know the length of the data in advance because you can read many DID in a single request and the length is never encoded.  

There is a non-standard hack you can do to read the whole payload. It only works for DID that are the last one being read in a request. This feature has been requested by a user, but it's nothing standard.  It can be achieved by throwing an exception in the len function.. not the cleanest implementation, but works for a hacky feature.  see `__len__` here : https://udsoncan.readthedocs.io/en/latest/udsoncan/helper_classes.html#udsoncan.DidCodec

**Comment by lapo4719** [comment]
@pylessard maybe don't support this idea.

I can think of a workaround and it probably isn't worth your time.
Right now I have a small list of Did codecs that I added manually, and then I built the ability to read a CDD file for the rest. My goal with this was to be able to provided a reduced ability to read DID data if a user doesn't provide a CDD file, or if the CDD file is missing a DID entry because of being out of date.

But instead of asking you to build a default mechanism, I can just build a table from 0x0000 to 0xFFFF and add my default in post for any DIDs that aren't assigned one yet. It's bulky, but you won't have to add anything.

I will pursue this, and you can choose if you'd like to support a default_codec to save space, but optionally

thank you very much for looking into this, and the DidCodec.ReadAllRemainingData suggestion

**Comment by pylessard** [comment]
ok
You can also design your own extension of a dict that behave like you'd like

```
class MyDefaultDict(dict):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.default_val = None
    
    def set_default(self, v):
        self.default_val = v
    
    def __contains__(self, k):
        return super().__contains__(k) or self.default_val is not None
    
    def __getitem__(self, k):
        if not super().__contains__(k) and self.default_val is not None:
            return self.default_val
        return super().__getitem__(k)
```

**Comment by lapo4719** [comment]
Tried this approach, it works to create the missing key: value pair, but this has unexpected consequences:
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/52b20984-5d80-4f8e-9a78-8c3f9d47f922)
The `if 'data_identifiers' in didconfig` causes that entry to be registered with the default value (a codec, not a dictionary) and then line 179 re-assigns the whole dictionary.

**Comment by lapo4719** [comment]
Regarding the ReadAllRemainingData trick. This does not appear to work for the DTCInformation interpret_response() specifically, because it needs to iterate over the remaining length, and doesn't yet handle this exception.

![image](https://github.com/pylessard/python-udsoncan/assets/48290683/c88045f9-fcaa-443d-8563-8ee943d172c0)
![image](https://github.com/pylessard/python-udsoncan/assets/48290683/1cd22481-49f4-4e5f-93b2-c7f26c75f520)

**Comment by pylessard** [comment]
I implemented the default codec config.
Want to try this branch?  

[support-default-did-codec](https://github.com/pylessard/python-udsoncan/tree/support-default-did-codec)

**Comment by pylessard** [comment]
I have merged the branch. reopen this issue if that doesn't fix your issue.
Fixed in #194

**Comment by lapo4719** [comment]
Thank you so much for taking to time to add this. I'll be happy to test this out and confirm. but I'm unable to till after Christmas.

**Comment by JensLukaschkowitz** [comment]
Dear @pylessard, 

in your merge request #194 you added a deep copy of the complete config structure in your check_did_config function in dids.py. This function is implicitly called twice while using Client.read_data_by_identifier by the make_request and interpret_response methods. So say a user user uses this read_data_by_identifier function per DID and has 20 DID, this will result in 40 deep copy operations, or? 

And if the request is integrated in a time-based loop, so per time step 2*n, with n is the number of DIDs, deepcopy operations will occur. For big datapoint dictionaries this is from my point of view very time consuming.

Is there a way to reduce the number of deep copy operations to one or to only make a deep copy for the default case? In my case by commenting the deepcopy line I was able to reduce the time per DID read from 2,5s to 0,15s on a PI Zero.

Best regards Jens

**Comment by pylessard** [comment]
Interesting, did not consider the performance hit caused by this. will review soon

**Comment by pylessard** [comment]
@JensLukaschkowitz   
Try this branch : https://github.com/pylessard/python-udsoncan/tree/fix-performance-degradation-caused-by-deepcopy-of-didconfig

Any better?

**Comment by JensLukaschkowitz** [comment]
@pylessard works great for me, thx for this quick fix

**Comment by pylessard** [comment]
Fixed in #205
